consensus/src/chained_bft/block_storage/block_tree.rs

```diff
@@ -249,23 +264,42 @@ where
         if let Some(old_qc) = self.id_to_quorum_cert.get(&block_id) {
             return VoteReceptionResult::OldQuorumCertificate(Arc::clone(old_qc));
         }
-        // All the votes collected for all the execution results of a given proposal.
-        let block_votes = self
-            .id_to_votes
-            .entry(block_id)
-            .or_insert_with(HashMap::new);
+
+        let author = vote_msg.author();
+        if let Some(old_block_id) = self.author_to_block.get(&author) {
+            if block_id == *old_block_id {
+                // Author has already voted for this block
+                return VoteReceptionResult::DuplicateVote;
+            }
+            if let Some(old_block_hash) = self.id_to_votes.get_mut(&old_block_id) {
+                let digest = old_block_hash.digest;
+                let old_block_votes = &mut old_block_hash.li_sig;
+                if let Some(li_sig) = old_block_votes.get_mut(&digest) {
+                    if li_sig.remove_signature(author) {
+                        // Last vote/signature for block removed, cleanup hashmap.
+                        old_block_votes.remove(&digest);
+                        self.id_to_votes.remove(&old_block_id);
```

Comment:
I don't think this is correct. You're allowed to remove the entry only if there are no signatures for other li_digests: the hash_map of digests to signatures should be empty.

Reply:
Hmm, entry is removed if and only if this was the last signature in list implying there was only one vote on this block. May be I am missing something obvious. Let's discuss tomorrow.

Reply:
I absolutely agree that we want to remove the entry with this specific digest. However, the same block id might keep votes with other LI digests.
Consider a case, in which we're aggregating honest votes for some `block_id` and `li_digest_good`. A malicious proposer sends you some dummy vote for `block_id` and `li_digest_bad`. At this moment you have two entries in the hashmap of the `block_id`. When the same malicious proposer sends another dummy vote for `other_block_id` we're going to cleanup the previous vote and remove `li_digest_bad`'s entry. However, we don't want to remove `block_id` entry because it still has the signatures corresponding to `li_digest_good`.
Does it make sense?